cmd/clusterctl/phases/pivot.go

```diff
@@ -0,0 +1,366 @@
+	}
+
+	// New objects cannot have a specified resource version. Clear it out.
+	cluster.SetResourceVersion("")
+	if err := to.CreateClusterObject(cluster); err != nil {
+		return errors.Wrapf(err, "error copying Cluster %s/%s to target cluster", cluster.Namespace, cluster.Name)
+	}
+
+	// TODO: Once there is a more formal link between Machine* objects and Clusters this will need to be
+	// updated, currently they are linked through existing in the same Namespace.
+	klog.V(4).Infof("Retrieving list of MachineDeployments to move for Cluster %s/%s", cluster.Namespace, cluster.Name)
+	machineDeployments, err := from.GetMachineDeploymentObjectsInNamespace(cluster.Namespace)
+	if err != nil {
+		return err
+	}
+	if err := moveMachineDeployments(from, to, machineDeployments); err != nil {
+		return err
+	}
+
+	klog.V(4).Infof("Retrieving list of MachineSets not associated with a MachineDeployment to move for Cluster %s/%s", cluster.Namespace, cluster.Name)
+	machineSets, err := from.GetMachineSetObjectsInNamespace(cluster.Namespace)
+	if err != nil {
+		return err
+	}
+	if err := moveMachineSets(from, to, machineSets); err != nil {
+		return err
+	}
+
+	klog.V(4).Infof("Retrieving list of Machines not associated with a MachineSet to move for Cluster %s/%s", cluster.Namespace, cluster.Name)
+	machines, err := from.GetMachineObjectsInNamespace(cluster.Namespace)
+	if err != nil {
+		return err
+	}
+	if err := moveMachines(from, to, machines); err != nil {
+		return err
+	}
+
+	if err := from.ForceDeleteCluster(cluster.Namespace, cluster.Name); err != nil {
+		return errors.Wrapf(err, "error force deleting cluster %s/%s", cluster.Namespace, cluster.Name)
+	}
+
+	klog.V(4).Infof("Successfully moved Cluster %s/%s", cluster.Namespace, cluster.Name)
+	return nil
+}
+
+func moveMachineDeployments(from, to clusterclient.Client, machineDeployments []*clusterv1.MachineDeployment) error {
+	machineDeploymentNames := make([]string, 0, len(machineDeployments))
+	for _, md := range machineDeployments {
+		machineDeploymentNames = append(machineDeploymentNames, md.Name)
+	}
+	klog.V(4).Infof("Preparing to move MachineDeployments: %v", machineDeploymentNames)
+
+	for _, md := range machineDeployments {
+		if err := moveMachineDeployment(from, to, md); err != nil {
+			return errors.Wrapf(err, "failed to move MachineDeployment %s:%s", md.Namespace, md.Name)
+		}
+	}
+	return nil
+}
+
+func moveMachineDeployment(from, to clusterclient.Client, md *clusterv1.MachineDeployment) error {
+	klog.V(4).Infof("Moving MachineDeployment %s/%s", md.Namespace, md.Name)
+	klog.V(4).Infof("Retrieving list of MachineSets for MachineDeployment %s/%s", md.Namespace, md.Name)
+	machineSets, err := from.GetMachineSetsForMachineDeployment(md)
+	if err != nil {
+		return err
+	}
+
+	// TODO: Moved machine deployments do not appear to be re-adopted properly by MachineDeployments
+	// At a minimum cascading delete does not work properly following the move
```

Comment:
Did you mean `Moved machine sets`?

Reply:
I was wondering given that MachineSets can readopt Machines, if we should instead just let the target cluster recreate the machineset and adopt the existing machines instead of pivoting the sets in a deployment.

Reply:
> I was wondering given that MachineSets can readopt Machines, if we should instead just let the target cluster recreate the machineset and adopt the existing machines instead of pivoting the sets in a deployment.

It would require testing to see if it would work, however from what I can tell this would only possibly work for the latest deployed version of the MachineDeployment and post-pivot any MachineSets associated with previous deployments would be lost. You would also lose some of the metadata that is stored in the annotations of the MachineSet with this approach.

Reply:
> Did you mean `Moved machine sets`?

Yes, but I'm going to remove this comment because there is a different comment discussing the workaround for managing owner refs and a link to the github issue.